src/utils/devtools.ts

```diff
@@ -29,61 +33,68 @@ export const devtools = <T extends object>(proxyObject: T, name?: string) => {
   const devtools = extension.connect({ name })
-  const unsub1 = subscribe(proxyObject, () => {
+  const unsub1 = subscribe(proxyObject, (ops) => {
+    const action = ops
+      .filter(([_, path]) => path[0] !== DEVTOOLS)
+      .flatMap(([op, path]) => `${op}:${path.map(String).join('.')}`)
+      .join(', ')
+
+    if (!action) {
+      return
+    }
+
     if (isTimeTraveling) {
       isTimeTraveling = false
     } else {
+      const proxyWithoutDevtools = Object.assign({}, snapshot(proxyObject))
+      delete (proxyWithoutDevtools as any)[DEVTOOLS]
+
       devtools.send(
-        `Update - ${new Date().toLocaleString()}`,
-        snapshot(proxyObject)
+        `${action} - ${new Date().toLocaleString()}`,
+        proxyWithoutDevtools
       )
     }
   })
-  const unsub2 = devtools.subscribe(
-    (message: { type: string; payload?: any; state?: any }) => {
-      if (message.type === 'DISPATCH' && message.state) {
-        if (
-          message.payload?.type === 'JUMP_TO_ACTION' ||
-          message.payload?.type === 'JUMP_TO_STATE'
-        ) {
-          isTimeTraveling = true
-        }
-        const nextValue = JSON.parse(message.state)
-        Object.keys(nextValue).forEach((key) => {
-          ;(proxyObject as any)[key] = nextValue[key]
-        })
-      } else if (
-        message.type === 'DISPATCH' &&
-        message.payload?.type === 'COMMIT'
-      ) {
-        devtools.init(snapshot(proxyObject))
-      } else if (
-        message.type === 'DISPATCH' &&
-        message.payload?.type === 'IMPORT_STATE'
+  const unsub2 = devtools.subscribe((message: Message) => {
+    if (message.type === 'DISPATCH' && message.state) {
+      if (
+        message.payload?.type === 'JUMP_TO_ACTION' ||
+        message.payload?.type === 'JUMP_TO_STATE'
       ) {
-        const actions = message.payload.nextLiftedState?.actionsById
-        const computedStates =
-          message.payload.nextLiftedState?.computedStates || []
-
         isTimeTraveling = true
+      }
+      ;(proxyObject as any)[DEVTOOLS] = message
+    } else if (
+      message.type === 'DISPATCH' &&
+      message.payload?.type === 'COMMIT'
+    ) {
+      devtools.init(snapshot(proxyObject))
+    } else if (
+      message.type === 'DISPATCH' &&
+      message.payload?.type === 'IMPORT_STATE'
+    ) {
+      const actions = message.payload.nextLiftedState?.actionsById
+      const computedStates =
+        message.payload.nextLiftedState?.computedStates || []
 
-        computedStates.forEach(({ state }: { state: any }, index: number) => {
-          const action =
-            actions[index] || `Update - ${new Date().toLocaleString()}`
+      isTimeTraveling = true
 
-          Object.keys(state).forEach((key) => {
-            ;(proxyObject as any)[key] = state[key]
-          })
+      computedStates.forEach(({ state }: { state: any }, index: number) => {
+        const action =
+          actions[index] || `Update - ${new Date().toLocaleString()}`
```

Comment:
Do you know when action is not found?
It feels like this is fine or it would reveal it.
```suggestion
          actions[index] || 'No action found'
```

Reply:
Yes, i think this is a better solution!